### ISSUE TITLE
Applying SSL to class-imbalanced data; how to change MSN loss to support arbitrary priors?

### ISSUE THREAD

**Issue by faris-k** [lightly-ai/lightly#1058]
This paper describes how certain self-supervised learning frameworks can benefit from adopting non-uniform priors for improved performance on class-imbalanced datasets: 

The Hidden Uniform Cluster Prior in Self-Supervised Learning
https://arxiv.org/abs/2210.07277
> Abstract: A successful paradigm in representation learning is to perform self-supervised pretraining using tasks based on mini-batch statistics (e.g., SimCLR, VICReg, SwAV, MSN). We show that in the formulation of all these methods is an overlooked prior to learn features that enable uniform clustering of the data. While this prior has led to remarkably semantic representations when pretraining on class-balanced data, such as ImageNet, we demonstrate that it can hamper performance when pretraining on class-imbalanced data. By moving away from conventional uniformity priors and instead preferring power-law distributed feature clusters, we show that one can improve the quality of the learned representations on real-world class-imbalanced datasets. To demonstrate this, we develop an extension of the Masked Siamese Networks (MSN) method to support the use of arbitrary features priors.

They propose an extension of MSN that supports the use of arbitrary feature priors, and it looks like they do so by just modifying the loss.

## Previous MSN loss:
![image](https://user-images.githubusercontent.com/68213622/217393278-6ebe5f59-5ff7-4399-8cd2-2dcfc47b2cec.png)

## New loss: Prior Matching for Siamese Networks, PMSN:
![image](https://user-images.githubusercontent.com/68213622/217393619-9aa13a8a-0c08-4564-9ebc-8d70a90081a6.png)

What would I need to adjust in [msn_loss.py](https://github.com/lightly-ai/lightly/blob/master/lightly/loss/msn_loss.py) to make the change from MSN loss to this PMSN loss? Would it be somewhere here, where mean entropy maximization regularization is performed?

https://github.com/lightly-ai/lightly/blob/7fcd165fd09b05df62eaa30c2a70f337d4113e1e/lightly/loss/msn_loss.py#L196-L201

Also, any general advice on applying self-supervised learning to class-imbalanced data?

**Comment by guarin** [comment]
Hi @faris-k, thanks for bringing this paper up! I already had my eyes on it for a while but didn't have the time to look more into it.

I think it should be relatively straight forward to implement. You can use the KL-loss from pytorch to calculate the extra loss: https://pytorch.org/docs/stable/generated/torch.nn.functional.kl_div.html and we already calculate the mean anchor probabilities $\bar{p}$ as `mean_anchor_probs = torch.mean(anchor_probs, dim=0)` here: https://github.com/lightly-ai/lightly/blob/7fcd165fd09b05df62eaa30c2a70f337d4113e1e/lightly/loss/msn_loss.py#L198

The only thing that remains is passing the target distribution. For this I would suggest adding the following arguments to the  `MSNLoss.__init__` method:
* `target_distribution: Union[str, Tensor] = "uniform"`
* `power_law_exponent: float = 0.25`
* `regularization_weight: float = 1.0`
* `me_max_weight: float | None = None` # deprecate old regularization weight

The `target_distribution` can either be set to `"uniform"`, `"power"`, or a custom tensor (must have same size as $p$). If it is `"uniform"` we apply the default MSN loss. If it is `"power"` we calculate a power law distribution using the `power_law_exponent` argument. If it is a tensor we use the tensor directly as target distribution.

The `regularization_weight` could act as the new lambda parameter. For MSN we used `me_max_weight` but we should be more general. So I would propose to add the new `regularization_weight` argument and set `me_max_weight` to `None` by default and mark it as deprecated in favor of `regularization_weight` in the docstring. If `me_max_weight` is not `None` we should keep using the uniform prior assumption and us `me_max_weight` for regularization as this means that someone is using old code to run the model. If `me_max_weight` is `None` and `target_distribution` is not `"uniform"` we should error and tell the user to use `regularization_weight` instead.

I hope this is enough to get you started :) Also note that you can skip the whole complications with the `me_max_weight` deprecation and supporting different distributions if you just want to implement the PMSN loss. I somehow wrote the response as a brain-dump on how I would implement it.

> Also, any general advice on applying self-supervised learning to class-imbalanced data?

I think the paper itself gives some great advice on this: Don't use methods that make assumptions about the class distribution. MAE is such a method and as Table 1 shows is relatively stable when it comes to different distributions.

<img width="774" alt="Screen Shot 2023-02-10 at 14 19 08" src="https://user-images.githubusercontent.com/43336610/218102080-7c07d018-10e9-4759-af0d-68f36b56dba5.png">

**Comment by yfchen1997** [comment]
Excuse me, is the loss calculated by kl divergence negative?

**Comment by guarin** [comment]
Hi :)

We use KLDivLoss from pytorch: https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html as you can see here: https://github.com/lightly-ai/lightly/blob/3ea0736139af1db01152f44393c394daca6c973e/lightly/loss/pmsn_loss.py#L71

The loss is not strictly negative, but I believe in practice it usually is.

**Comment by yfchen1997** [comment]
Thank you reply. When i use pmsn loss, i find the regularization loss = F.kl_div(input=mean_anchor_probs, target=power_dist, reduction="sum") is not changed. And cross-entropy loss = torch.mean(torch.sum(torch.log(probs**(-targets)), dim=1)) is decreasing.
Have you ever encountered such a situation?
Epoch 0 - Epoch 13: regularization loss = -6.89. cross-entropy loss=5.6->3.5.

**Comment by guarin** [comment]
I haven't encountered this. Maybe you could check if the `mean_anchor_probs` are changing?

**Comment by yfchen1997** [comment]
mean_anchor_probs  are changing, but why can power_dist remain unchanged? The power_dist value obtained by the following calculation formula remains unchanged.
            size = mean_anchor_probs.shape[0]
            exponent = self.power_law_exponent
            k = torch.arange(1, size + 1).cuda()
            power_dist = k ** (-exponent)
            power_dist = power_dist / power_dist.sum()
            rloss = F.kl_div(input=mean_anchor_probs, target=power_dist, reduction="sum")

**Comment by guarin** [comment]
Power dist is the target distribution, it should remain unchanged. For normal MSN loss the target distribution is uniform which also remains unchanged over the run.

**Comment by yfchen1997** [comment]
13 epochs has been trained, mean_anchor_probs changes slightly, but rloss remains unchanged.
<img width="684" alt="截屏2023-05-10 下午4 27 54" src="https://github.com/lightly-ai/lightly/assets/52944433/8323153c-eeca-42dc-b583-0fb8467812d5">

**Comment by yfchen1997** [comment]
And  when i use me_max_loss, all together seem normal.
